mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -292,34 +350,111 @@ impl SaplingSyncLoopHandle {
+                    debug!("Got block {:?}", block);
+                    let mut compact_txs = Vec::new();
+                    // By default, CompactBlocks only contain CompactTxs for transactions that contain Sapling spends or outputs.
+                    // Create and push compact_tx during iteration.
+                    for (tx_id, hash_tx) in block.tx.iter().enumerate() {
+                        let tx_bytes = client.get_transaction_bytes(hash_tx).compat().await?;
+                        let tx = ZTransaction::read(tx_bytes.as_slice()).unwrap();
+                        let mut spends = Vec::new();
+                        let mut outputs = Vec::new();
+                        if !tx.shielded_spends.is_empty() || !tx.shielded_outputs.is_empty() {
+                            // Create and push spends with outs for compact_tx during iterations.
+                            for spend in &tx.shielded_spends {
+                                let compact_spend = TonicCompactSpend {
+                                    nf: spend.nullifier.to_vec(),
+                                };
+                                spends.push(compact_spend);
+                            }
+                            for out in &tx.shielded_outputs {
+                                let compact_out = TonicCompactOutput {
+                                    cmu: out.cmu.to_bytes().to_vec(),
+                                    epk: out.ephemeral_key.to_bytes().to_vec(),
+                                    // https://zips.z.cash/zip-0307#output-compression
+                                    // The first 52 bytes of the ciphertext contain the contents and opening of the note commitment,
+                                    // which is all of the data needed to spend the note and to verify that the note is spendable.
+                                    ciphertext: out.enc_ciphertext[0..52].to_vec(),
+                                };
+                                outputs.push(compact_out);
+                            }
+                            // Shadowing mut variables as immutable. No longer need to update them.
+                            let spends = spends;
+                            let outputs = outputs;
+                            let mut hash_tx_vec = hash_tx.0.to_vec();
+                            hash_tx_vec.reverse();
+
+                            let compact_tx = TonicCompactTx {
+                                index: tx_id as u64,
+                                hash: hash_tx_vec,
+                                fee: 0,
+                                spends,
+                                outputs,
+                            };
+                            compact_txs.push(compact_tx);
+                        }
+                    }
+                    let mut hash = block.hash.0.to_vec();
+                    hash.reverse();
+                    let mut prev_hash = block.previousblockhash.unwrap().0.to_vec();
```

Comment:
Using `.unwrap` here might lead to an occasional panic. When `previousblockhash` can be `None`?

Reply:
Yeah, you are right. Genesis block has no previous block, so it’s previous hash value set to 0.
I used unwrap_or_default()